airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java

```diff
@@ -138,24 +138,75 @@ private PreparedStatement createXminQueryStatement(
+      return connection.prepareStatement(sql);
+    } else if (isSingleWraparound(prevRunXminStatus, currentXminStatus)) {
       // The xmin state that we save represents the lowest XID that is still in progress. To make sure we
-      // don't miss
-      // data associated with the current transaction, we have to issue an >=
+      // don't miss data associated with the current transaction, we have to issue an >=. Because of the wraparound, the changes prior to the
+      // end xmin xid value must also be captured.
+      final String sql = String.format("SELECT %s FROM %s WHERE xmin::text::bigint >= ? OR xmin::text::bigint < ?",
+          wrappedColumnNames, fullTableName);
+      final PreparedStatement preparedStatement = connection.prepareStatement(sql);
+      if (prevRunXminStatus != null) {
+        preparedStatement.setLong(1, prevRunXminStatus.getXminXidValue());
+      } else {
+        preparedStatement.setLong(1, 0L);
+      }
+
+      preparedStatement.setLong(2, currentXminStatus.getXminXidValue());
+      return preparedStatement;
+    } else {
+      // The xmin state that we save represents the lowest XID that is still in progress. To make sure we
+      // don't miss data associated with the current transaction, we have to issue an >=
       final String sql = String.format("SELECT %s FROM %s WHERE xmin::text::bigint >= ?",
           wrappedColumnNames, fullTableName);
 
       final PreparedStatement preparedStatement = connection.prepareStatement(sql.toString());
 
-      final XminStatus currentStreamXminStatus = xminStateManager.getXminStatus(airbyteStream);
-      if (currentStreamXminStatus != null) {
-        preparedStatement.setLong(1, currentStreamXminStatus.getXminXidValue());
+      if (prevRunXminStatus != null) {
+        preparedStatement.setLong(1, prevRunXminStatus.getXminXidValue());
       } else {
         preparedStatement.setLong(1, 0L);
       }
-      LOGGER.info("Executing query for table {}: {}", tableName, preparedStatement);
+
       return preparedStatement;
-    } catch (final SQLException e) {
-      throw new RuntimeException(e);
+    }
+  }
+
+  // Detect whether the source Postgres DB has undergone a single wraparound event.
+  static boolean isSingleWraparound(final XminStatus prevRunXminStatus, final XminStatus currentXminStatus) {
+    if (prevRunXminStatus != null) {
+      return currentXminStatus.getNumWraparound() - prevRunXminStatus.getNumWraparound() == 1;
+    } else {
+      return currentXminStatus.getNumWraparound() - 0 == 1;
+    }
+  }
+
+  // Detects whether source Postgres DB has undergone multiple wraparound events between syncs.
+  static boolean shouldPerformFullSync(final XminStatus prevRunXminStatus, final XminStatus currentXminStatus) {
+    if (prevRunXminStatus != null) {
+      return currentXminStatus.getNumWraparound() - prevRunXminStatus.getNumWraparound() >= 2;
+    } else {
+      return currentXminStatus.getNumWraparound() - 0 >= 2;
```

Comment:
`- 0`?

Reply:
Refactored this to check for `null` previous state upfront. Please take a look